Ensure templates filenames without path and extension.

kaisermann opened the pull request:
I was reading the solution (#1939) to fix #1936 and came up with an alternative fix by changing the 

```php
    return collect($templates)
        ->map(function ($template) {
            return preg_replace('#\.(blade\.)?php$#', '', ltrim($template));
        })
...
```

with 

```php
    return collect($templates)
        ->map(function ($template) {
            return basename(basename(ltrim($template), '.php'), '.blade');
        })
...
```

QWp6t commented:
I think that might screw up custom page templates, and it can mess with any other hooks into template hierarchy that looks for the template in a subfolder, like certain plugins.

#1936 happens because WordPress uses [`get_page_template_slug()`](https://developer.wordpress.org/reference/functions/get_page_template_slug/) to get the path to the template file. It's mostly a cosmetic issue, but it's one that we can easily solve.

The solution presented here can introduce more bugs so I'm hesitant to accept it.

kaisermann commented:
Hm, good point! I'm gonna test it with subfolders and some plugins that offer template files and write the feedback here.

kaisermann commented:
Another approach, without the `basename`:

1) Modified regex `'#\.(blade\.)?php$#'` with `'#\.(blade\.?)?(php)?$#'` to catch `template-custom.blade` (which comes without the `.php`).

2) Moved `$paths` out of the flatMap callback

3) Made the flatMap check if a template name starts with each `$path` and remove it from the the string.

```php
function filter_templates($templates)
{
    $paths = apply_filters('sage/filter_templates/paths', [
        'views',
        'resources/views'
    ]);

    return collect($templates)
        ->map(function ($template) {
            return preg_replace('#\.(blade\.?)?(php)?$#', '', ltrim($template));
        })
        ->flatMap(function ($template) use ($paths) {
            return collect($paths)
                ->flatMap(function ($path) use ($template) {
                    if (strpos($template, '/')) {
                        $template = preg_replace("#^$path/#", '', $template);
                    }
                    return [
                        "{$path}/{$template}.blade.php",
                        "{$path}/{$template}.php",
                        "{$template}.blade.php",
                        "{$template}.php",
                    ];
                });
        })
        ->filter()
        ->all();
}
```

9585999 commented:
@kaisermann Nice 👍

with this https://github.com/soberwp/controller/pull/39 It works as it should.

kaisermann commented:
Being a little bit picky with the returned paths:

The previous code would still return some entries with inexistent file paths. 

Note the `$paths_pattern`.

```php
function filter_templates($templates)
{
    $paths = apply_filters('sage/filter_templates/paths', [
        'views',
        'resources/views'
    ]);
    $paths_pattern = "#^(" . implode('|', $paths) . ")/#";

    return collect($templates)
        ->map(function ($template) {
            return preg_replace('#\.(blade\.?)?(php)?$#', '', ltrim($template));
        })
        ->flatMap(function ($template) use ($paths, $paths_pattern) {
            if (strpos($template, '/')) {
                $template = preg_replace($paths_pattern, '', $template);
            }
            return collect($paths)
                ->flatMap(function ($path) use ($template) {
                    return [
                        "{$path}/{$template}.blade.php",
                        "{$path}/{$template}.php",
                        "{$template}.blade.php",
                        "{$template}.php",
                    ];
                });
        })
        ->unique()
        ->filter()
        ->all();
}
```

retlehs commented:
@kaisermann thanks! could you please push that code to this branch? then we'll get this merged in :+1:

kaisermann commented:
@retlehs Done 👍 !!
